nemo/core/classes/module.py

```diff
@@ -54,18 +55,59 @@ def input_example(self, max_batch=None, max_dim=None):
     def freeze(self) -> None:
         r"""
         Freeze all params for inference.
+
+        This method sets `requires_grad` to False for all parameters of the module.
+        It also stores the original `requires_grad` state of each parameter in a dictionary,
+        so that `unfreeze()` can restore the original state if `partial=True` is set in `unfreeze()`.
         """
-        for param in self.parameters():
+        grad_map = {}
+
+        for pname, param in self.named_parameters():
+            # Store the original grad state
+            grad_map[pname] = param.requires_grad
+            # Freeze the parameter
             param.requires_grad = False
 
+        # Store the frozen grad map
+        if not hasattr(self, '_frozen_grad_map'):
+            self._frozen_grad_map = grad_map
+        else:
+            self._frozen_grad_map.update(grad_map)
+
         self.eval()
 
-    def unfreeze(self) -> None:
+    def unfreeze(self, partial: bool = False) -> None:
```

Comment:
Regarding the name `partial`: my understanding is that this is intended as `only_adapters=False` but you don't want to name it as such to maintain separation of concerns between adapter and freeze/unfreeze APIs. I think the current name is a bit confusing when read out of the context of this PR; maybe we could at least extend the doc to explain why/when you might want to use this option?